riscv/macros/src/lib.rs

```diff
@@ -0,0 +1,366 @@
+            "PriorityNumber" => Ok(Self::Priority),
+            "HartIdNumber" => Ok(Self::HartId),
+            _ => Err(syn::Error::new(
+                trait_name.span(),
+                "Unknown trait name. Expected: 'ExceptionNumber', 'CoreInterruptNumber', 'ExternalInterruptNumber', 'PriorityNumber', or 'HartIdNumber'",
+            )),
+        }
+    }
+}
+
+/// Marker traits for interrupts
+enum InterruptType {
+    Core,
+    External,
+}
+
+impl InterruptType {
+    /// Returns a token stream representing the name of the marker trait
+    fn marker_trait_name(&self) -> TokenStream2 {
+        match self {
+            Self::Core => quote!(CoreInterruptNumber),
+            Self::External => quote!(ExternalInterruptNumber),
+        }
+    }
+
+    /// Returns a token stream representing the name of the array of interrupt service routines
+    fn isr_array_name(&self) -> TokenStream2 {
+        match self {
+            Self::Core => quote!(__CORE_INTERRUPTS),
+            Self::External => quote!(__EXTERNAL_INTERRUPTS),
+        }
+    }
+
+    /// Returns a token stream representing the name of the interrupt dispatch function
+    fn dispatch_fn_name(&self) -> TokenStream2 {
+        match self {
+            Self::Core => quote!(_dispatch_core_interrupt),
+            Self::External => quote!(_dispatch_external_interrupt),
+        }
+    }
+}
+
+/// Struct containing the information needed to implement the `riscv-pac` traits for an enum
+struct PacEnumItem {
+    /// The name of the enum
+    name: Ident,
+    /// The maximum discriminant value
+    max_number: usize,
+    /// A map from discriminant values to variant names
+    numbers: HashMap<usize, Ident>,
+}
+
+impl PacEnumItem {
+    fn new(input: &DeriveInput) -> Self {
+        let name = input.ident.clone();
+        let (mut numbers, mut max_number) = (HashMap::new(), 0);
+
+        let variants = match &input.data {
+            Data::Enum(data) => &data.variants,
+            _ => panic!("Input is not an enum"),
+        };
+        for v in variants.iter() {
+            let ident = v.ident.clone();
+            let value = match &v.discriminant {
+                Some(d) => match &d.1 {
+                    syn::Expr::Lit(expr_lit) => match &expr_lit.lit {
+                        syn::Lit::Int(lit_int) => match lit_int.base10_parse::<usize>() {
+                            Ok(num) => num,
+                            Err(_) => {
+                                panic!("All variant discriminants must be unsigned integers")
+                            }
+                        },
+                        _ => panic!("All variant discriminants must be unsigned integers"),
+                    },
+                    _ => panic!("All variant discriminants must be unsigned integers"),
+                },
+                _ => panic!("Variant must have a discriminant"),
+            };
```

Comment:
```suggestion
           let value = match v.discriminant.as_ref() {
                Some((_, syn::Expr::Lit(expr_lit))) => match &expr_lit.lit {
                    syn::Lit::Int(lit_int) => {
                        lit_int.base10_parse::<usize>().unwrap_or_else(|_| {
                            panic!("All variant discriminants must be unsigned integers")
                        })
                    }
                    _ => panic!("All variant discriminants must be unsigned integers"),
                },
                None => panic!("Variant must have a discriminant"),
                _ => panic!("All variant discriminants must be literal expressions"),
            };
```

Removes some of the indent creep.